# SQL
<br><br>
**Описание проекта**

*Цели*: Написать запросы для ответов на поставленные вопросы.

*Входные данные*: база данных (структура БД представлена на картинке).
<br><br>
**План проекта**:
- Создано подключение;
- Сделаны запросы и получены ответы на вопросы.

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#1.-Подключение-к-базе-данных" data-toc-modified-id="1.-Подключение-к-базе-данных-1">1. Подключение к базе данных</a></span></li><li><span><a href="#2.-Вывод-первых-строк-таблиц" data-toc-modified-id="2.-Вывод-первых-строк-таблиц-2">2. Вывод первых строк таблиц</a></span></li><li><span><a href="#3.-Ответы-на-вопросы" data-toc-modified-id="3.-Ответы-на-вопросы-3">3. Ответы на вопросы</a></span><ul class="toc-item"><li><span><a href="#3.1.-Количество-книг,-опубликованных-после-1-января-2000-года" data-toc-modified-id="3.1.-Количество-книг,-опубликованных-после-1-января-2000-года-3.1">3.1. Количество книг, опубликованных после 1 января 2000 года</a></span></li><li><span><a href="#3.2.-Количество-обзоров-и-средняя-оценка-книг" data-toc-modified-id="3.2.-Количество-обзоров-и-средняя-оценка-книг-3.2">3.2. Количество обзоров и средняя оценка книг</a></span></li><li><span><a href="#3.3.-Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="3.3.-Издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-3.3">3.3. Издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#3.4.-Автор-с-самой-высокой-средней-оценкой-книг-—-учитывать-только-книги-с-50-и-более-оценками" data-toc-modified-id="3.4.-Автор-с-самой-высокой-средней-оценкой-книг-—-учитывать-только-книги-с-50-и-более-оценками-3.4">3.4. Автор с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками</a></span></li><li><span><a href="#3.5.-Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="3.5.-Среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-3.5">3.5. Среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li></ul></div>

## 1. Подключение к базе данных

In [1]:
# <импорт библиотек>
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# <установка параметров>
db_config = {'user': '***', # <имя пользователя>
 'pwd': '***', # <пароль>
 'host': '***',
 'port': ***, # <порт подключения>
 'db': '***'} # <название базы данных>

# <создание строки подключения>
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# <сохранение коннектора>
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## 2. Вывод первых строк таблиц

**Таблица books**

In [3]:
# <вывод первых десяти строк таблицы books>
query = \
'''
SELECT *
FROM books
LIMIT 10
'''
pd.io.sql.read_sql(query, con = engine) 

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

- *book_id* - primary key;
- *author_id*, *publisher_id* - foreign keys.

**Таблица authors**

In [4]:
# <вывод первых десяти строк таблицы authors>
query = \
'''
SELECT *
FROM authors
LIMIT 10
'''
pd.io.sql.read_sql(query, con = engine) 

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

- *author_id* - primary key.

**Таблица ratings**

In [5]:
# <вывод первых десяти строк таблицы ratings>
query = \
'''
SELECT *
FROM ratings
LIMIT 10
'''
pd.io.sql.read_sql(query, con = engine) 

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

- *rating_id* - primary key;
- *book_id* - foreign key.

**Таблица reviews**

In [6]:
# <вывод первых десяти строк таблицы reviews>
query = \
'''
SELECT *
FROM reviews
LIMIT 10
'''
pd.io.sql.read_sql(query, con = engine) 

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

- *review_id* - primary key;
- *book_id* - foreign key.

## 3. Ответы на вопросы

### 3.1. Количество книг, опубликованных после 1 января 2000 года

In [7]:
# <количество книг, опубликованных после 1 января 2000 года>
query = \
'''
/* вывод количества книг, опубликованных после 1 января 2000 года */
SELECT COUNT(publication_date) AS books_count
FROM books
WHERE publication_date > '2000-01-01'
'''
pd.io.sql.read_sql(query, con = engine) 

books_count
0          819

**Выводы**

- После 1 января 2000 года было опубликовано **819 книг**;
- Книги, выпущенные 1 января 2000 года, в вычислениях не учитывались.

### 3.2. Количество обзоров и средняя оценка книг

In [8]:
# <количество обзоров и средняя оценка книг>
query = \
'''
/* вывод book_id, названия, количества обзоров и среднего рейтинга по книгам;
   для части книг обзоров нет, для них заменили null на 0;
   такую же проверку для рейтинга проводить не стали, так как рейтинг может быть нулевым,
   если же его нет, то будет стоять null */
SELECT b.book_id, b.title,
       CASE WHEN subq1.review_count IS NOT NULL THEN subq1.review_count ELSE 0 END AS review_count, 
       subq2.avg_rating
FROM books b LEFT JOIN (SELECT book_id, COUNT(book_id) AS review_count FROM reviews GROUP BY book_id) subq1
                       ON b.book_id=subq1.book_id
             LEFT JOIN (SELECT book_id, AVG(rating) AS avg_rating FROM ratings GROUP BY book_id) subq2
                       ON b.book_id=subq2.book_id
ORDER BY book_id
'''
reviews_and_rating = pd.io.sql.read_sql(query, con = engine)
reviews_and_rating

book_id                                              title  review_count  \
0          1                                       'Salem's Lot             2   
1          2                 1 000 Places to See Before You Die             1   
2          3  13 Little Blue Envelopes (Little Blue Envelope...             3   
3          4  1491: New Revelations of the Americas Before C...             2   
4          5                                               1776             4   
..       ...                                                ...           ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)             3   
996      997                        Xenocide (Ender's Saga  #3)             3   
997      998                                    Year of Wonders             4   
998      999                        You Suck (A Love Story  #2)             2   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...             4   

     avg_rating  
0      3.666667  
1      2.500000  
2      4.666667  
3      4.500000  
4      4.000000  
..          ...  
995    3.666667  
996    3.400000  
997    3.200000  
998    4.500000  
999    3.833333  

[1000 rows x 4 columns]

**Выводы**

- Всего в таблицу присутствует 1000 книг;
- При формировании запроса выяснилось, что у части книг нет обзоров, поэтому была применена конструкция *CASE WHEN* для замены значений *NULL* на *0*.

### 3.3. Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [9]:
# <издательство, которое выпустило наибольшее число книг толще 50 страниц>
query = \
'''
/* подсчет количества книг толще 50 страниц для каждого издательства */
WITH books_over50_counter AS
(SELECT publisher_id, COUNT(publisher_id) AS book_counter
FROM books 
WHERE num_pages > 50
GROUP BY publisher_id)
        
/* соединение таблицы с издательствами и таблицы с количеством книг,
   вывод издательства с максимальным количеством книг толще 50 страниц */
SELECT p.publisher_id, p.publisher, boc.book_counter
FROM publishers p INNER JOIN books_over50_counter boc ON p.publisher_id=boc.publisher_id
WHERE boc.book_counter = (SELECT MAX(book_counter) FROM books_over50_counter)
'''
pd.io.sql.read_sql(query, con = engine) 

publisher_id      publisher  book_counter
0           212  Penguin Books            42

**Выводы**

- Издательство, выпустившее наибольшее число книг толще 50 страниц - **Penguin Books** -> **42 книги**;
- Чтобы учесть случай, когда несколько издательств выпустило одинаковое число книг, была использована конструкция после *WHERE* в конце запроса вместо ключевого слова *LIMIT*.

### 3.4. Автор с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками

In [10]:
# <автор с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками>
# <расчет среднего среди всех книг без группировки по книгам>
query = \
'''
/* выражение для определения книг с количеством оценок >= 50 */
WITH ratings_over50 AS
(SELECT book_id, rating
FROM ratings
WHERE book_id IN (SELECT book_id FROM ratings GROUP BY book_id HAVING COUNT(book_id) >= 50)),
/* выражение для подсчета средней оценки по автору */
avg_author_ratings as
(SELECT b.author_id, AVG(ro50.rating) AS avg_rating
FROM books b INNER JOIN ratings_over50 ro50 ON b.book_id=ro50.book_id
GROUP BY b.author_id)
        
/* соединение с таблицей authors, выбор автора с максимальной оценкой книг */
SELECT a.author, aar.avg_rating
FROM avg_author_ratings aar INNER JOIN authors a ON aar.author_id=a.author_id
WHERE aar.avg_rating = (SELECT MAX(avg_rating) FROM avg_author_ratings)
'''
pd.io.sql.read_sql(query, con = engine)


author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

In [11]:
# <автор с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками>
# <расчет среднего среди средних оценок по книгам>
query = \
'''
/* выражение для определения средней оценки книг с количеством оценок >= 50 */
WITH ratings_over50 AS
(SELECT book_id, AVG(rating) AS book_avg_rating
FROM ratings
WHERE book_id IN (SELECT book_id FROM ratings GROUP BY book_id HAVING COUNT(book_id) >= 50)
GROUP BY book_id),
        
/* выражение для подсчета средней оценки по автору */
avg_author_ratings AS
(SELECT b.author_id, AVG(ro50.book_avg_rating) AS avg_rating
FROM books b INNER JOIN ratings_over50 ro50 ON b.book_id=ro50.book_id
GROUP BY b.author_id)
        
/* соединение с таблицей authors, выбор автора с максимальной оценкой книг */
SELECT a.author, aar.avg_rating
FROM avg_author_ratings aar INNER JOIN authors a ON aar.author_id=a.author_id
WHERE aar.avg_rating = (SELECT MAX(avg_rating) FROM avg_author_ratings)
'''
pd.io.sql.read_sql(query, con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

**Выводы**

- Авторы с самой высокой оценкой книг (при расчете среднего среди всех оценок без группировки по книгам) - **J.K. Rowling/Mary GrandPré** -> средняя оценка составляет **4.287** из 5;
- Авторы с самой высокой оценкой книг (при расчете среднего среди средних оценок книг) - также **J.K. Rowling/Mary GrandPré** -> средняя оценка чуть ниже и составляет **4.284** из 5.

### 3.5. Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [12]:
# <среднее количество обзоров от пользователей, которые поставили больше 50 оценок>
query = \
'''
/* выбор пользователей, которые поставили более 50 оценок */
WITH users_over50_ratings AS
(SELECT username
FROM ratings
GROUP BY username
HAVING count(username) > 50),
/* подсчет количества обзоров от пользователей, которые поставили больше 50 оценок */
users_over50_ratings_reviews AS
(SELECT username, COUNT(review_id) AS cnt_reviews
FROM reviews
WHERE username IN (SELECT username FROM users_over50_ratings)
GROUP BY username)

/* вывод среднего количества обзоров от пользователей, которые поставили более 50 оценок */
SELECT AVG(cnt_reviews) AS avg_cnt_reviews
FROM users_over50_ratings_reviews
'''
pd.io.sql.read_sql(query, con = engine)

avg_cnt_reviews
0        24.333333

**Выводы**

- В среднем пользователи, которые поставили более 50 оценок, также делают около **24.3 обзоров**.